---
title: "monaLisa - MOtif aNAlysis with Lisa"
author: "Dania Machlab, Lukas Burger, Michael Stadler"
date: "`r Sys.Date()`"
bibliography: monaLisa-refs.bib
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.width = 7,
    fig.height = 5,
    out.width = "80%",
    fig.align = "center",
    crop = NULL # suppress "The magick package is required to crop" issue
)
```

<img src="monaLisa_logo_v1.png" style="display: block; margin-left: auto; margin-right: auto; width: 50%; border: 0" />

# Introduction

`monaLisa` is a collection of functions for working with biological sequences and motifs
that represent the binding preferences of transcription factors or nucleic acid binding proteins.

For example, `monaLisa` can make use of her father [Homer](http://homer.ucsd.edu/homer/index.html)
to look for enriched motifs in sets of genomic regions, compared to all other regions.

The motifs come from a collection of transcription factor binding site specificities,
such as `JASPAR2018`. The regions could be for example promoters, non-promoters (distal regions)
or accessible regions. The regions are grouped into bins according to a numerical value
assigned to each bin, such as change of expression or accessibility.

`monaLisa` can also be used to look for motifs in sequences.

Finally, `monaLisa` can can be used to predict what transcription factors are likely
to explain observed changes in chromatin accessibility between two conditions.



# Quick example: Identify enriched motifs

```{r, quick, eval=FALSE}
# load package
library(monaLisa)

# bin regions
# (atac.peaks.change is a numerical vector)
bins <- bin(x = atac.peaks.change, binmode = "equalN", nElement = 400)

# dump motifs into file
dumpJaspar("jaspar2018.motif", pkg = "JASPAR2018")

# find Homer (findMotifsGenome.pl)
homerfile <- findHomer(dirs = "/path/to/look/into")

# run analysis
# (atac.peaks is a GRanges)
se <- runHomer(gr = atac.peaks, b = bins, genomedir = "/work/gbioinfo/DB/genomes/mm10",
               outdir = "myresults", motifFile = "jaspar2018.motif", homerfile = homerfile,
               regionsize = "given", Ncpu = 4L)

```

`se` is a `SummarizedExperiment` with `assays` *p*, *FDR*, *enr* and *log2enr*, each containing a matrix
with motifs (rows) by bins (columns). The values are:  

- *p*: the raw P value ($-\log_{10} p$) of a given motif enrichment in a given bin,
  as reported by Homer. Each P value results from a Homer analysis testing each
  motif occuring within the bin, compared to its occurrences in all other bins
  as a background.  
- *FDR*: Same as *p* but adjusted for multiple testing using the method of
  Benjamini and Hochberg, 1995 (`p.adjust(..., method="fdr"`).  
- *enr*: Motif enrichments, calculated as: $enr = (o - e) / \sqrt e$, where $o$ and
  $e$ are the observed and expected numbers of regions in the bin containing a
  given motif. These enrichments correspond to $z$ values ($z = (z - \mu)/\sigma$),
  assuming that the numbers of regions in a bin that contain a motif are independent
  Poisson random variables with mean and variance $e$.
- *log2enr*: Motif enrichments, calculated as: $log2enr = log2((o + 8)/(e + 8))$,
  where $o$ and $e$ are the observed and expected numbers of regions in the bin
  containing a given motif.

In addition, `rowData(se)` and `colData(se)` give information about the used motifs
and bins, respectively. In `metadata(se)` you can find information about parameter
values.

# Detailed example: Finding TFs enriched in differentially methylated regions

The detailed example is based on an in vitro differentiation system, in which
mouse embryonic stem (ES) cells are differentiated into neuronal progenitors (NP).
In an earlier study [@LMRs], we have analyzed the genome-wide CpG methylation patterns
in these cell types and identified so call *low methylated regions* (LMRs), that
have reduced methylation levels and correspond to regions bound by transcription
factors.

We have later published a tool that systematically identifies such regions from
genome-wide methylation data [@MethylSeekR]. Interestingly, a change in methylation
of LMRs is indicative of altered transcription factor binding. We will therefore
use these regions to identify TF motifs that are enriched/depleted in LMR regions
that change their methylation between ES and NP cell states.


## Load library
We start by loading the needed libraries:  
```{r loadlib, message=FALSE}
library(GenomicRanges)
library(SummarizedExperiment)
library(monaLisa)
```

## Genomic regions of interest
`monaLisa` contains a file with genomic coordinates (mouse mm9 assembly) of LMRs,
with the respective changes of methylation:  
```{r loadLMRs}
lmrfile <- system.file("extdata", "LMRsESNPmerged.gr.rds", package = "monaLisa")
lmr <- readRDS(lmrfile)
lmr
```

You can see there are `r length(lmr)` LMRs, most of which gain methylation between
ES and NP stages:  
```{r deltameth}
hist(lmr$deltaMeth, 100, col = "gray", main = "",
     xlab = "Change of methylation (NP - ES)", ylab = "Number of LMRs")
```

In order to keep the computation time reasonable, we'll select 10,000 of the LMRs randomly:  
```{r lmrsel}
set.seed(1)
lmrsel <- lmr[ sample(x = length(lmr), size = 10000, replace = FALSE) ]
```


## Bin genomic regions
Now let's bin our LMRs by how much they change methylation, using the `bin` function.
We are not interested in small changes of methylation, say less than 0.3, so we'll
use the `minAbsX` argument to create a *no-change* bin in [-0.3, 0.3). The remaining
LMRs are put into bins of 800 each:  
```{r binlmrs}
bins <- bin(x = lmrsel$deltaMeth, binmode = "equalN", nElement = 800, minAbsX = 0.3)
table(bins)
```

<!--
saveRDS(bins, "/work/gbioinfo/stadler/development/R/compbio-lisa/inst/extdata/bins.rds")
-->

Because of the asymmetry of methylation changes, there is only a single bin with LMRs
that lost methylation and many that gained:  
```{r plotbins}
plotBinDensity(lmrsel$deltaMeth, bins, legend = "topleft")
```

Note that the bin-breaks around the *no-change* bin are not exactly -0.3 to 0.3.
They have been adjusted to have the required 800 LMRs per bin.

## Prepare motif enrichment analysis
Next we prepare the Homer-based motif enrichment analysis. We first need to create
a file with the motifs that we are interested in. A simple way is to dump the vertebrate
motifs from the `JASPAR2018` package into a file:  
```{r dumpjaspar}
motiffile <- tempfile(fileext = ".motif")
dumpJaspar(motiffile, pkg = "JASPAR2018")
```

Next we need to know where Homer is installed (you may have to adjust the `dirs`
argument below according to your environment):  
```{r homerscript}
homerfile <- findHomer(dirs = "/work/gbioinfo/Appz/Homer/Homer-4.11/bin/")
```

## Run motif enrichment analysis
Finally, we run the analysis. Again, you may have to adjust the `genomedir`.

This step requires that you have Homer installed on your system. Also, it may
take a while, for this reason you do not need to run it, but can just load the
results as shown below:  
```{r runhomer, eval=FALSE}
outdir <- tempfile(fileext = ".output")
se <- runHomer(gr = lmrsel, b = bins, genomedir = "/work/gbioinfo/DB/genomes/mm9",
               outdir = outdir, motifFile = motiffile, homerfile = homerfile,
               regionsize = "given", Ncpu = 20L)
```

<!--
saveRDS(se, "/tungstenfs/groups/gbioinfo/stadler/development/R/monaLisa/inst/extdata/se.rds")
-->

In case you did not run the above code, let's now read in the results:  
```{r gethomerresults}
se <- readRDS(system.file("extdata", "se.rds", package = "monaLisa"))
```

`se` is a `SummarizedExperiment` object which nicely keeps motifs, bins and
corresponding metadata together:  
```{r summarizedexperiment}
# summary
se
dim(se) # motifs-by-bins

# motif info
rowData(se)
head(rownames(se))

# bin info
colData(se)
head(colnames(se))

# assays: the motif enrichment results
assayNames(se)
assay(se, "log2enr")[1:5, 1:3]
```

We can plot the results using the `plotMotifHeatmaps` function, e.g. selecting all factors that have an absolute log2 enrichment of at least 1.0 in any bin:  
```{r plottfs}
# select strongly enriched TFs
sel <- apply(assay(se, "log2enr"), 1, function(x) max(abs(x))) > 1.0
sum(sel)
seSel <- se[sel, ]
# shorten names
rownames(seSel) <- sub("\\_.*$","",rownames(seSel))
# plot
plotMotifHeatmaps(x = seSel, which.plots = c("log2enr","FDR"), width = 2.0,
                  cluster = TRUE, maxEnr = 2, maxSig = 10, show_motif_GC = TRUE)
```

As the Jaspar database contains many similar motifs, it is helpful to show the enrichment heatmap clustered by motif similarity. To this end, we first need to calculate all pairwise motif similarities (measured as maximum Pearson correlation of all possible shifted alignments). As this step may take a while, you can also skip this line of code and load the results using the next line of code. 

```{r wmclustering, eval=FALSE}
SimMatSel <- motifSimilarity(rowData(seSel)$motif.pfm, BPPARAM = BiocParallel::MulticoreParam(20))
```

<!--
saveRDS(SimMatSel, "/tungstenfs/groups/gbioinfo/stadler/development/R/monaLisa/inst/extdata/SimMatSel.rds")
-->

```{r getclusteringresults}
SimMatSel <- readRDS(system.file("extdata", "SimMatSel.rds", package = "monaLisa"))
```
The order of the TFs in the resulting matrix is consistent with the elements of `seSel`.
Using this matrix, we can create an *hclust* object for clustering and plot the clustered heatmap. The plotting of the dendogram is controlled by the parameter *show_dendrogram*. 
```{r plottfsclustered}
# create hclust object, similarity defined by 1 - Pearson correlation
hcl <- hclust(as.dist(1 - SimMatSel), method = "average")
plotMotifHeatmaps(x = seSel, which.plots = c("log2enr","FDR"), width = 1.2,
                  cluster = hcl, maxEnr = 2, maxSig = 10,
                  show_dendrogram = TRUE, show_seqlogo = TRUE,
                  width.seqlogo = 1.2)
```


# Use `monaLisa` to annotate genomic regions with predicted motifs

As mentioned `monaLisa` can also be used to scan sequences for motifs. Here is an
example (just on few sequences/motifs for illustration):

```{r findMotifs}
# get sequences of promoters as a DNAStringSet
# (could also be a single DNAString, or the name of a fasta file)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
gr <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, upstream = 1000, downstream = 500)[c(1,4,5,10)]
library(BSgenome.Hsapiens.UCSC.hg19)
seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg19, gr)
seqs

# get motifs as a PWMatrixList
# (could also be a single PWMatrix, or the name of a motif file)
library(JASPAR2018)
library(TFBSTools)
pwms <- getMatrixSet(JASPAR2018, list(matrixtype = "PWM", tax_group = "vertebrates"))
pwms <- pwms[c("MA0885.1","MA0099.3","MA0033.2","MA0037.3","MA0158.1")]
pwms

# predict hits in sequences
res <- findMotifHits(query = pwms,
                     subject = seqs,
                     min.score = 6.0,
                     method = "matchPWM",
                     BPPARAM = BiocParallel::MulticoreParam(2))
res

# ... or using method = "homer2"
homerfile <- findHomer(homerfile = "homer2", dirs = "/work/gbioinfo/Appz/Homer/Homer-4.11/bin/")
if (!is.na(homerfile)) {
    # use homer2 if available
    res2 <- findMotifHits(query = pwms, subject = seqs, min.score = 6.0,
                          method = "homer2", homerfile = homerfile)
} else {
    # otherwise, use matchPWM
    res2 <- findMotifHits(query = pwms, subject = seqs, min.score = 6.0,
                          method = "matchPWM")
}
res2

summary(res %in% res2)

# create hit matrix:
# number of site of each motif per sequence
m <- table(seqnames(res), as.character(res$pwmname))
m
```



# Motif Selection with Randomized Lasso Stability Selection

One can also select for meaningful transcription factors (TFs) in a regression based approach with predictors consisting of the TFs, and the response vector being the observed log-fold changes across a set of regions. In `monaLisa`, we implement the randomized lasso stability selection proposed by Meinshausen and Bühlmann [@StabSel] with the improved error bounds introduced by Shah and Samworth [@compStabs]. We have modified the `stabsel` function from the `stabs` package to implement the randomized lasso version of stability selection. 

Lasso stability selection implements the lasso regression multiple times on subsamples of the data, and returns a selection probability for each predictor (number of times selected divided by number of regressions done). With the randomized lasso, a weakness parameter is additionally used to randomly vary the lambda parameter of the lasso regression. Although the main appeal of stability selection is in cases where the number of predictors exceeds the number of observations, it also performs better variable selection with noisy data [@StabSel].

With this approach, all TFs compete against each other to explain all observations, with the flexibility of adding other things like GC content to compete against the TFs for selection as well.

In the example below, we select for TFs explaining log-fold changes in chromatin accessibility (ATAC-seq) across the enhancers between mouse liver and lung tissue at P0, but this can be applied to other data types as well (ChIP-seq, RNA-seq, methylation etc.).

<!--

Selecting Transcription Factors Likely to Explain Changes in Chromatin Accessibility Using Stability Selection

Randomized Lasso stability selection proposed by Meinshausen and Bühlmann [@StabSel] applies the lasso regression on several subsamples of the data using a weakness parameter that adds additional variation to the lambda parameter and allows for more correlated variables to be selected together. Although the main appeal of stability selection is that it allows for selection of predictors when the number of predictors exceeds the number of observations, the authors also argue for this as a way of slightly overcoming the problem of lasso only selecting variables that are not correlated, and performing better in a noisy dataset depending on how strongly one sets the weakness parameter. Shah and Samworth [@compStabs] have further improved error control for stability selection using complementary pairs stability selection. The R `stabs` package [@stabs] implements the Shah and Samworth version of stability selection. `monaLisa` uses a modified version of the `stabsel` function in `stabs` that performs randomized lasso stability selection.

Here we demonstrate the use of randomized lasso stability selection to select transcription factors (TFs) that are likely to explain changes in chromatin accessibility from ATAC-seq data. One can also use stability selection on any other type of data like ChIP-seq of specific histone marks for example. We are given a vector of log-fold changes (logFC) in chromatin accessibility between two conditions that we consider to be our response vector. The predictor matrix will consist of the number of transcription factor binding sites (TFBS) that each TF has across the peaks. The example data set comes from ENCODE [@encode] and consists of bam files for ATAC-seq from mouse postnatal (P0) liver (identifiers: ENCFF146ZCO, ENCFF109LQF) and lung (identifiers: ENCFF203DOC, ENCFF823PTD) tissues. Each condition had 2 replicates, and the peak calling was done per condition using MACS2 [@macs]. The peaks from both conditions were merged and used to calculate ATAC-seq counts for all samples with QuasR [@QuasR]. Counts per million (CPM) were calculated per sample and averaged across replicates. The logFC was calculated as [log2(CPM(Liver)+8) - log2(CPM(Lung)+8)].

Only distal autosomal peaks (distance > 1kb from any TSS) were kept since promoter peaks tend not to change so much in accessibility, and we are interested in selecting TFs that likely explain logFC in enhancer regions. The `.rds` files used in monaLisa's example dataset consist of peaks that had an absolute logFC greater than 2 and less than 2.2. This was done to reduce the number of peaks and speed up runtime to demonstrate the functions used in `monaLisa` that apply randomized lasso stability selection [@StabSel].

-->

## Load Data

We load the ATAC-seq log2-fold changes across the enhancers (ATAC peaks). This will be our response vector in the regression framework.

```{r libs}
suppressPackageStartupMessages({
  library(monaLisa)
  library(JASPAR2018)
  library(TFBSTools)
  library("BSgenome.Mmusculus.UCSC.mm10")
  library(Biostrings)
  library(SummarizedExperiment)
  library(ComplexHeatmap) # heatmap drawing
  library(circlize) # used for color specification
})
```


```{r loadData}
# load logFC vector and peaks
peaks_path <- system.file("extdata", "Liver_vs_Lung_ATAC_peaks.rds", package = "monaLisa")
response_path <- system.file("extdata", "Liver_vs_Lung_ATAC_logFC.rds", package = "monaLisa")
logFC <- readRDS(response_path)
peaks <- readRDS(peaks_path)

# examine data
head(logFC)
head(peaks)
```

## Get TFBS per Motif Across the Peaks

We will now construct the transcription factor binding site (TFBS) matrix for known motifs (from a database like JASPAR) across the peak regions. We use the `findMotifHits` function to get the TF hits. This matrix will be the predictor matrix in our regression.

As mentioned, this framework offers the flexibility to add additional predictors to compete against the TFs for selection. Here, we add the fraction of G+C and CpG observed/expected ratio as predictors, to ensure that selected TFs are not just detecting a simple trend in GC or CpG composition.

```{r predictorMatrix}
# get PWMs (random subset of 150 vertebrate TFs from Jaspar)
pfms <- getMatrixSet(JASPAR2018, list(matrixtype = "PFM", tax_group = "vertebrates")) # for seq-logos
set.seed(6698)
pfms <- pfms[sort(sample(length(pfms), size = 150), decreasing = FALSE)]
pwms <- toPWM(pfms) # for searching



# get TFBS on given GRanges (peaks)
peakSeq <- BSgenome::getSeq(BSgenome.Mmusculus.UCSC.mm10, peaks)
hits <- findMotifHits(query = pwms, subject = peakSeq, min.score = 6.0,
                      method = "matchPWM",
                      BPPARAM = BiocParallel::MulticoreParam(2))

# get TFBS matrix
TFBSmatrix <- unclass(table(factor(seqnames(hits), levels = seqlevels(hits)),
                                  factor(hits$pwmname, levels = TFBSTools::name(pwms))))
TFBSmatrix[1:6, 1:6]

# check order
all(rownames(TFBSmatrix) == names(peaks))
all(rownames(TFBSmatrix) == names(logFC))

# calculate GC and oeCpG content
fMono <- oligonucleotideFrequency(peakSeq, width = 1L, as.prob = TRUE)
fDi <- oligonucleotideFrequency(peakSeq, width = 2L, as.prob = TRUE)
fracGC <- fMono[,"G"] + fMono[,"C"]
oeCpG <- (fDi[,"CG"] + 0.01) / (fMono[,"G"] * fMono[,"C"] + 0.01)

# add GC and oeCpG to predictor matrix
TFBSmatrix <- cbind(oeCpG, TFBSmatrix)
TFBSmatrix[1:6, 1:6]
```


## Identify important TFs

We can now run randomized lasso stability selection to identify TFs that are likely to explain the log-fold changes in accessibility. 

```{r stabSelTFs}
# select TFs
set.seed(123)
se <- randLassoStabSel(x = TFBSmatrix, y = logFC,
                       cutoff = 0.8, mc.cores = 2)
se

# selected TFs
colnames(se)[se$selected]
```

<!--
saveRDS(predictor_matrix, "/tungstenfs/groups/gbioinfo/stadler/development/R/monaLisa/inst/extdata/stabs_predictor.rds")
saveRDS(response, "/tungstenfs/groups/gbioinfo/stadler/development/R/monaLisa/inst/extdata/stabs_response.rds")
saveRDS(stabs, "/tungstenfs/groups/gbioinfo/stadler/development/R/monaLisa/inst/extdata/stabs.rds")
-->


```{r plotStabilityPaths}
# plot stability paths 
plotStabilityPaths(se)
```

Each line corresponds to a TF, and we can see the selection probabilities as a function of the regularization step (lambda parameter in lasso). The TF selection happens at the last step, given the specified minimum probability.

We can also visualize the selection probabilities of the selected TFs, optionally multiplied by the sign of the correlation to the logFC vector (response vector), to know how the TF relates to the logFC (`directional` parameter).

```{r plotSelProbs, fig.width=15, fig.height=7}
plotSelectionProb(se, directional = TRUE, selProbMinPlot = 0.6)
```


Next, we visualize the correlation structure of the TFBS matrix. We can see that the selected TFs have a similar correlation structure as the whole TFBS matrix.

```{r TFBScor}
# correlation matrix (excluding oeCpG and fracGC)
excl <- colnames(TFBSmatrix) %in% c("oeCpG", "fracGC")
TFBSmatrixCor <- cor(TFBSmatrix[, !excl], method = "pearson")
sel <- colnames(se)[se$selected]
TFBSmatrixCorSel <- TFBSmatrixCor[sel, sel]

# heatmap
Heatmap(TFBSmatrixCor,
        show_row_names = FALSE, show_column_names = FALSE, 
        name = "Pear. Cor.", column_title = "All TFs",
        col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red")),
        use_raster = TRUE)

pfmsSel <- pfms[match(colnames(TFBSmatrixCorSel), name(pfms))]
maxwidth <- max(sapply(TFBSTools::Matrix(pfmsSel), ncol))
seqlogoGrobs <- lapply(pfmsSel, seqLogoGrob, xmax = maxwidth)
hmSeqlogo <- rowAnnotation(logo = anno_seqlogo(seqlogoGrobs, which = "row"),
                           annotation_width = unit(1.5, "inch"),
                           show_legend = FALSE, show_annotation_name = FALSE)

Heatmap(TFBSmatrixCorSel, show_row_names = TRUE, show_column_names = TRUE, 
        name = "Pear. Cor.", column_title = "Selected TFs",
        col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red")), 
        right_annotation = hmSeqlogo)
```

We can examine the peaks that have hits for a selected TF of interest, ordered by absolute logFC values.

```{r topPeaks}
TF <- sel[2]
TF

i <- which(assay(se, "x")[, TF] > 0) # peaks that contain TF hits...
nm <- names(sort(abs(logFC[i]), decreasing = TRUE)) # ... order by |logFC|

head(peaks[nm])
```


# Session info and logo
The monaLisa logo uses a drawing that was obtained from http://vectorish.com/lisa-simpson.html
under the Creative Commons attribution - non-commercial 3.0 license: https://creativecommons.org/licenses/by-nc/3.0/.

This vignette was built using:  
```{r, session}
sessionInfo()
```

# References
