---
title: "monaLisa - MOtif aNAlysis with Lisa"
author: "Dania Machlab, Lukas Burger, Michael Stadler"
date: "`r Sys.Date()`"
bibliography: monaLisa-refs.bib
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{monaLisa - MOtif aNAlysis with Lisa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.width = 7,
    fig.height = 5,
    out.width = "80%",
    fig.align = "center",
    crop = NULL # suppress "The magick package is required to crop" issue
)
library(BiocStyle)
```

<img src="monaLisa_logo_v1.png" style="display: block; margin-left: auto; margin-right: auto; width: 50%; border: 0" />


# Introduction

`monaLisa` is a collection of functions for working with biological sequences
and motifs that represent the binding preferences of transcription factors or
nucleic acid binding proteins.

For example, `monaLisa` can be used to conveniently find motif hits in
sequences (see section \@ref(findhits)), or to identify motifs that are likely
associated to observed experimental data. This analysis is supposed to
provide potential answers to the question "Which transcription factors are
the drivers of my observed changes in expression/methylation/accessibility?".

You can use known motifs from collections of transcription factor binding 
specificities such as `JASPAR2018`, also available from 
[Bioconductor](https://bioconductor.org/packages/JASPAR2018/). Genomic
regions could be for example promoters, enhancers or accessible regions for
which experimental data is available.

Two independent approaches are implemented to identify interesting motifs:

- In **binned motif enrichment analysis** (`calcBinnedMotifEnr()`, see
section \@ref(binned)), genomic regions are grouped into bins according to a
numerical value assigned to each region, such as change of expression,
accessibility or methylation. Motif enrichments are then calculated for
each bin, normalizing for differences in sequence composition very similarly
as it was originally done by [Homer](http://homer.ucsd.edu/homer/index.html).

- In **randomized Lasso stability selection** (`randLassoStabSel()`, see
section \@ref(stabsel)), `monaLisa` uses a robust regression approach to predict
what transcription factors can explain experimental measurements,
for example changes in chromatin accessibility between two conditions. Also
this approach allows to correct for sequence composition. In addition, similar
motifs have to "compete" with each other to be selected.

For both approaches, functions that allow visualization of obtained results
are provided.

If you prefer to jump right in, you can continue with section \@ref(quick) that
shows a quick hypothetical example of how to run a binned motif enrichment
analysis. If you prefer to actually compute enrichments on real data, you can
find below detailed examples for a binned motif enrichment analysis (section
\@ref(binned)) and randomized Lasso stability selection (section \@ref(stabsel)).


# Quick example: Identify enriched motifs in bins{#quick}

```{r, quick, eval=FALSE}
# load package
library(monaLisa)

# bin regions
# (peak_change is a numerical vector)
peak_bins <- bin(x = peak_change, binmode = "equalN", nElement = 400)

# calculate motif enrichments
# (peak_seqs is a DNAStringSet, pwms is a PWMatrixList)
se <- calcBinnedMotifEnr(seqs = peak_seqs,
                         bins = peak_bins,
                         motifs = pwms)
```

`se` is a `SummarizedExperiment` with `assays` *negLog10P*, *negLog10Padj*, *pearsonResid*, *log2enr*, *sumForegroundWgtWithHits* and *sumBackgroundWgtWithHits*, each containing a matrix
with motifs (rows) by bins (columns). The values are:  

- *negLog10P*: the raw P value ($-\log_{10} p$) of a given motif enrichment in a given
  bin. Each P value results from an enrichment calculation comparing occurrences
  of each motif in the bin to its occurrences in background sequences, defined by
  the `background` argument (by default: sequences in all other bins).  
- *negLog10Padj*: Same as *negLog10P* but adjusted for multiple testing using the method of
  provided in the `p.adjust.method` argument, by default: Benjamini and Hochberg, 1995
  (`p.adjust(..., method="fdr")`).  
- *pearsonResid*: Pearson residuals, a measure for motif enrichments calculated
  as: $enr = (o - e) / \sqrt e$, where $o$ and $e$ are the observed and expected
  numbers of regions in the bin containing a given motif. These enrichments
  correspond to $z$ values ($z = (o - \mu)/\sigma$), assuming that the numbers of
  regions in a bin that contain a motif are independent Poisson random variables
  with mean and variance $e$.
- *log2enr*: Motif enrichments, calculated as: $log2enr = log2((o + c)/(e + c))$,
  where $o$ and $e$ are the observed and expected numbers of regions in the bin
  containing a given motif, and $c$ is a pseudocount defined by the
  `pseudocount.log2enr` argument.  
- *sumForegroundWgtWithHits* and *sumBackgroundWgtWithHits* are the sum of
  foreground and background sequences that have at least one occurrence of the
  motif, respectively. The background sequences are weighted in order to
  adjust for differences in sequence composition between foreground and
  background.

In addition, `rowData(se)` and `colData(se)` give information about the used motifs
and bins, respectively. In `metadata(se)` you can find information about parameter
values.


# Binned motif enrichment analysis: Finding TFs enriched in differentially methylated regions{#binned}

This example is based on an *in vitro* differentiation system, in which
mouse embryonic stem (ES) cells are differentiated into neuronal progenitors (NP).
In an earlier study [@LMRs], we have analyzed the genome-wide CpG methylation patterns
in these cell types and identified so called *low methylated regions* (LMRs), that
have reduced methylation levels and correspond to regions bound by transcription
factors.

We also developed a tool that systematically identifies such regions from
genome-wide methylation data [@MethylSeekR]. Interestingly, a change in methylation
of LMRs is indicative of altered transcription factor binding. We will therefore
use these regions to identify transcription factor motifs that are enriched
or depleted in LMRs that change their methylation between ES and NP cell
states.


## Load packages
We start by loading the needed packages:  
```{r loadlib, message=FALSE}
library(GenomicRanges)
library(SummarizedExperiment)
library(JASPAR2018)
library(TFBSTools)
library(BSgenome.Mmusculus.UCSC.mm10)
library(monaLisa)
```


## Genomic regions of interest
`monaLisa` contains a file with genomic coordinates (mouse mm10 assembly) of LMRs,
with the respective changes of methylation:  
```{r loadLMRs}
lmrfile <- system.file("extdata", "LMRsESNPmerged.gr.rds", package = "monaLisa")
lmr <- readRDS(lmrfile)
lmr
```

You can see there are `r length(lmr)` LMRs, most of which gain methylation between
ES and NP stages:  
```{r deltameth}
hist(lmr$deltaMeth, 100, col = "gray", main = "",
     xlab = "Change of methylation (NP - ES)", ylab = "Number of LMRs")
```

In order to keep the computation time reasonable, we'll select 10,000 of the LMRs randomly:  
```{r lmrsel}
set.seed(1)
lmrsel <- lmr[ sample(x = length(lmr), size = 10000, replace = FALSE) ]
```


## Bin genomic regions
Now let's bin our LMRs by how much they change methylation, using the `bin` function.
We are not interested in small changes of methylation, say less than 0.3, so we'll
use the `minAbsX` argument to create a *no-change* bin in [-0.3, 0.3). The remaining
LMRs are put into bins of 800 each:  
```{r binlmrs}
bins <- bin(x = lmrsel$deltaMeth, binmode = "equalN", nElement = 800, minAbsX = 0.3)
table(bins)
```


We can see which bin has been set to be the zero bin using `monaLisa::getZeroBin`, 
or set it to a different bin using `monaLisa::setZeroBin`:
```{r}
# find the index of the level representing the zero bin 
levels(bins)
getZeroBin(bins)
```


Because of the asymmetry of methylation changes, there is only a single bin with LMRs
that lost methylation and many that gained:  
```{r plotbins}
plotBinDensity(lmrsel$deltaMeth, bins, legend = "topleft")
```

Note that the bin-breaks around the *no-change* bin are not exactly -0.3 to 0.3.
They have been adjusted to have the required 800 LMRs per bin.


## Prepare motif enrichment analysis
Next we prepare the motif enrichment analysis. We first need known motifs
representing transcription factor binding site preferences. We extract all
vertebrate motifs from the `r Biocpkg("JASPAR2018")` package as positional
weight matrices (PWMs):  
```{r getmotifs}
pwms <- getMatrixSet(JASPAR2018,
                     opts = list(matrixtype = "PWM",
                                 tax_group = "vertebrates"))
```

Furthermore, we need the sequences corresponding to our LMRs, which
we directly extract from the `r Biocpkg("BSgenome.Mmusculus.UCSC.mm10")`
package:  
```{r getseqs}
lmrseqs <- getSeq(BSgenome.Mmusculus.UCSC.mm10, lmrsel)
```


## Run motif enrichment analysis
Finally, we run the binned motif enrichment analysis.

This step will take a while, and typically you would use the `BPPARAM`
argument to run it with parallelization. For this example however, you can just
load the pre-computed results as shown below.  
```{r runbinned, eval=FALSE}
se <- calcBinnedMotifEnr(seqs = lmrseqs, bins = bins, motifs = pwms)
```

In case you did not run the above code, let's now read in the results:  
```{r gethomerresults}
se <- readRDS(system.file("extdata", "se.rds", package = "monaLisa"))
```

`se` is a `SummarizedExperiment` object which nicely keeps motifs, bins and
corresponding metadata together:  
```{r summarizedexperiment}
# summary
se
dim(se) # motifs-by-bins

# motif info
rowData(se)
head(rownames(se))

# bin info
colData(se)
head(colnames(se))

# assays: the motif enrichment results
assayNames(se)
assay(se, "log2enr")[1:5, 1:3]
```

We can plot the results using the `plotMotifHeatmaps` function, e.g. selecting all
transcription factor motifs that have an $-log_{10} FDR$ of at least 8.0 in any bin
(corresponding to an $FDR < 10^{-8}$). FDR values are stored in the `negLog10Padj` assay:  
```{r plottfs}
# select strongly enriched motifs
sel <- apply(assay(se, "negLog10Padj"), 1, function(x) max(abs(x), 0, na.rm = TRUE)) > 8.0
sum(sel)
seSel <- se[sel, ]

# plot
plotMotifHeatmaps(x = seSel, which.plots = c("log2enr", "negLog10Padj"), width = 2.0,
                  cluster = TRUE, maxEnr = 2, maxSig = 10, show_motif_GC = TRUE)
```

Setting `cluster = TRUE` has re-ordered the rows using hierarchical clustering
of the `"pearsonResid"` assay. As many transcription factor binding motifs are similar to
each other, it is also helpful to show the enrichment heatmap clustered by
motif similarity. To this end, we first calculate all pairwise motif similarities
(measured as maximum Pearson correlation of all possible shifted alignments).
This can be quickly calculated for the few selected motifs using the `motifSimilarity`
function. For many motifs, this step may take a while, and it may be useful to
parallelize it using the `BPPARAM` argument (e.g. to run on `n` parallel threads
using the multi-core backend, you can use:
`motifSimilarity(..., BPPARAM = BiocParallel::MulticoreParam(n))`).

```{r wmclustering}
SimMatSel <- motifSimilarity(rowData(seSel)$motif.pfm)
range(SimMatSel)
```

The order of the TFs in the resulting matrix is consistent with the elements of
`seSel`, and the maximal similarity between any pair of motifs is 1.0. By
subtracting these similarities from 1.0, we obtain distances that we use
to perform a hierarchical clustering with the `hclust` function. The returned
object is then passed to the `cluster` argument to define the order of the rows
in the heatmap. The plotting of the dendrogram is controlled by the parameter
`show_dendrogram`, and we also display the motifs as sequence logos using
`show_seqlogo`: 
```{r plottfsclustered}
# create hclust object, similarity defined by 1 - Pearson correlation
hcl <- hclust(as.dist(1 - SimMatSel), method = "average")
plotMotifHeatmaps(x = seSel, which.plots = c("log2enr", "negLog10Padj"), width = 1.8,
                  cluster = hcl, maxEnr = 2, maxSig = 10,
                  show_dendrogram = TRUE, show_seqlogo = TRUE,
                  width.seqlogo = 1.2)
```


## Convert between motif text file for `Homer` and motif objects in `R`

`calcBinnedMotifEnr` is a wrapper function for `calcBinnedMotifEnrR` and `calcBinnedMotifEnrHomer`. `calcBinnedMotifEnrR` implements the binned motif enrichment analysis in `R`, similarly to `Homer` and does not require the user to have the `Homer` tool pre-installed. For more information on that function and how it resembles the `Homer` tool see the function documentation.

A simple way to represent a DNA sequence motif that assumes independence of positions in the motif is a matrix with four rows (for the bases A, C, G and T) and `n` columns for the `n` positions in the motif. The values in that matrix can represent the sequence preferences of a binding protein in several different ways:  

- **Position frequency matrices (PFM)** contain values that correspond to the number of times (frequency) that a given base has been observed in at a given position of the motif. It is usually obtained from a set of known, aligned binding site sequences, and depending on the number of sequences, the values will be lower or higher. In `R`, PFMs are often represented using `TFBSTools::PFMatrix` (single motif) or `TFBSTools::PFMatrixList` objects (set of motifs). This is the rawest way to represent a sequence motif and can be converted into any other representation.  
- **Position probability matrices (PPM)** are obtained by dividing the counts in each column of a PFM by their sum. The values now give a probability of observing a given base at that position of the motif and sum up to one in each column. This is the representation used in motif text files for `Homer`. A PPM can only be converted back to a PFM by knowing or assuming how many binding site sequences were observed (see argument `n` in `homerToPFMatrixList`).  
- **Position weight matrices (PWM)** (also known as position specific scoring matrices, PSSM) are obtained by comparing the base probabilities in a PPM to the probabilities of observing each base outside of a binding site (background base probabilities), for example by calculating log-odds scores (see `?toPWM` for details). This is a useful representation for scanning sequences for motif matches. In `R`, PWMs are often represented using `TFBSTools::PWMatrix` (single motif) or `TFBSTools::PWMatrixList` (set of motifs).

`calcBinnedMotifEnrR` takes PWMs as a `PWMatrixList` object to scan for motif hits. `calcBinnedMotifEnrHomer` on the other hand takes a motif text file with PPMs, and requires the user to have `Homer` installed to use it for the binned motif enrichment analysis. Here, we show how one can get motif PFMs from `JASPAR2018` and convert them to a `Homer`-compatible text file with PPMs (`dumpJaspar`) and vice versa (`homerToPFMatrixList` ), and how to convert a `PFMatrixList` to a `PWMatrixList` for use with `calcBinnedMotifEnrR` or `findMotifHits`: 

```{r}
# get PFMs from JASPAR2018 package (vertebrate subset)
pfms <- getMatrixSet(JASPAR2018,
                     opts = list(matrixtype = "PFM",
                                 tax_group = "vertebrates"))

# convert PFMs to PWMs
pwms <- toPWM(pfms)

# convert JASPAR2018 PFMs (vertebrate subset) to Homer motif file
tmp <- tempfile()
convert <- dumpJaspar(filename = tmp,
                      pkg = "JASPAR2018",
                      pseudocount = 0,
                      opts = list(tax_group = "vertebrates"))

# convert Homer motif file to PFMatrixList
pfms_ret <- homerToPFMatrixList(filename = tmp, n = 100L)

# compare the first PFM
# - notice the different magnitude of counts (controlled by `n`)
# - notice that with the default (recommended) value of `pseudocount = 1.0`,
#   there would be no zero values in pfms_ret matrices, making
#   pfms and pfms_ret even more different
as.matrix(pfms[[1]])
as.matrix(pfms_ret[[1]])

# compare position probability matrices with the original PFM 
# (also adding a pseudo-count of 1.0 first)
round(sweep(x = as.matrix(pfms[[1]]) + 1, MARGIN = 2, 
            STATS = colSums(as.matrix(pfms[[1]]) + 1), FUN = "/"), 3)
round(sweep(x = as.matrix(pfms_ret[[1]]), MARGIN = 2, 
            STATS = colSums(as.matrix(pfms_ret[[1]])), FUN = "/"), 3)
```


# Use `monaLisa` to annotate genomic regions with predicted motifs{#findhits}

As mentioned `monaLisa` can also be used to scan sequences for motifs. Here is an
example (just on a few sequences/motifs for illustration):

```{r findMotifs}
# get sequences of promoters as a DNAStringSet
# (could also be a single DNAString, or the name of a fasta file)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
gr <- promoters(TxDb.Mmusculus.UCSC.mm10.knownGene,
                upstream = 1000, downstream = 500)[c(1, 4, 5, 10)]
library(BSgenome.Mmusculus.UCSC.mm10)
seqs <- getSeq(BSgenome.Mmusculus.UCSC.mm10, gr)
seqs

# get motifs as a PWMatrixList
# (could also be a single PWMatrix, or the name of a motif file)
library(JASPAR2018)
library(TFBSTools)
pfms <- getMatrixByID(JASPAR2018, c("MA0885.1", "MA0099.3", "MA0033.2", "MA0037.3", "MA0158.1"))
pwms <- toPWM(pfms)
pwms
name(pwms)

# predict hits in sequences
res <- findMotifHits(query = pwms,
                     subject = seqs,
                     min.score = 6.0,
                     method = "matchPWM",
                     BPPARAM = BiocParallel::MulticoreParam(2))
res

# create hit matrix:
# number of sites of each motif per sequence
m <- table(factor(seqnames(res), levels = names(seqs)),
           factor(res$pwmname, levels = name(pwms)))
m
```

The transformation of sequence and PWM names to factors with defined levels
in the creation of the hit matrix above is not strictly needed, but it ensures
that even sequences or motifs without any hits are reported in the matrix,
and that the order of sequences (rows) and motifs (columns) is identical to
the order in `seqs` and `pwms`.


# Motif selection with Randomized Lasso Stability Selection{#stabsel}

One can also select for important transcription factor (TF) motifs in a regression based approach with predictors consisting of the TF motif hits, and the response vector being the observed measure of interest, e.g. log-fold changes of accessibility for a set of regions. In `monaLisa`, we implement the randomized lasso stability selection proposed by @StabSel with the improved error bounds introduced by @compStabs. We have modified the `glmnet.lasso` function used by `stabsel` from the `r CRANpkg("stabs")` package to implement the randomized lasso.

Lasso stability selection implements the lasso regression multiple times on subsamples of the data, and returns a selection probability for each predictor (number of times selected divided by number of regressions done). With the randomized lasso, a weakness parameter is additionally used  to vary the lasso penalty term $\lambda$ to a randomly chosen value between [$\lambda$, $\lambda$/weakness] for each predictor. Although the main appeal of stability selection is in cases where the number of predictors exceeds the number of observations, it also performs better variable selection with noisy data [@StabSel].

With this approach, TF motifs compete against each other to explain the response, and we can also include additional predictors like GC content to compete against the TF motifs for selection. This is especially useful if the response is biased by sequence composition, for example if regions with higher GC content tend to have higher response values.

In the example below, we select for TF motifs explaining log-fold changes in chromatin accessibility (ATAC-seq) across the enhancers between mouse liver and lung tissue at P0, but this can be applied to other data types as well (ChIP-seq, RNA-seq, methylation etc.).

<!--

Selecting Transcription Factors Likely to Explain Changes in Chromatin Accessibility Using Stability Selection

Randomized Lasso stability selection proposed by Meinshausen and Bühlmann [@StabSel] applies the lasso regression on several subsamples of the data using a weakness parameter that adds additional variation to the lambda parameter and allows for more correlated variables to be selected together. Although the main appeal of stability selection is that it allows for selection of predictors when the number of predictors exceeds the number of observations, the authors also argue for this as a way of slightly overcoming the problem of lasso only selecting variables that are not correlated, and performing better in a noisy dataset depending on how strongly one sets the weakness parameter. Shah and Samworth [@compStabs] have further improved error control for stability selection using complementary pairs stability selection. The R `stabs` package [@stabs] implements the Shah and Samworth version of stability selection. `monaLisa` uses a modified version of the `stabsel` function in `stabs` that performs randomized lasso stability selection.

Here we demonstrate the use of randomized lasso stability selection to select transcription factors (TFs) that are likely to explain changes in chromatin accessibility from ATAC-seq data. One can also use stability selection on any other type of data like ChIP-seq of specific histone marks for example. We are given a vector of log-fold changes (logFC) in chromatin accessibility between two conditions that we consider to be our response vector. The predictor matrix will consist of the number of transcription factor binding sites (TFBS) that each TF has across the peaks. The example data set comes from ENCODE [@encode] and consists of bam files for ATAC-seq from mouse postnatal (P0) liver (identifiers: ENCFF146ZCO, ENCFF109LQF) and lung (identifiers: ENCFF203DOC, ENCFF823PTD) tissues. Each condition had 2 replicates, and the peak calling was done per condition using MACS2 [@macs]. The peaks from both conditions were merged and used to calculate ATAC-seq counts for all samples with QuasR [@QuasR]. Counts per million (CPM) were calculated per sample and averaged across replicates. The logFC was calculated as [log2(CPM(Liver)+8) - log2(CPM(Lung)+8)].

Only distal autosomal peaks (distance > 1kb from any TSS) were kept since promoter peaks tend not to change so much in accessibility, and we are interested in selecting TFs that likely explain logFC in enhancer regions. The `.rds` files used in monaLisa's example dataset consist of peaks that had an absolute logFC greater than 2 and less than 2.2. This was done to reduce the number of peaks and speed up runtime to demonstrate the functions used in `monaLisa` that apply randomized lasso stability selection [@StabSel].

-->

## Load packages

We start by loading the needed packages: 
```{r libs, message=FALSE}
library(monaLisa)
library(JASPAR2018)
library(TFBSTools)
library(BSgenome.Mmusculus.UCSC.mm10)
library(Biostrings)
library(SummarizedExperiment)
library(ComplexHeatmap)
library(circlize)
```

## Load dataset

In our example dataset, we have quantified ATAC-seq reads on enhancers in mouse P0 lung and liver tissues. The log2-fold change (our response vector in this example) is for liver vs lung chromatin accessibility. We are using a set of 10,000 randomly sampled enhancers to illustrate how randomized lasso stability selection can be used to select TFs.
```{r loadData}
# load GRanges object with logFC and peaks
gr_path <- system.file("extdata", "atac_liver_vs_lung.rds", package = "monaLisa")
gr <- readRDS(gr_path)
```


## Get TFBS per motif and peak

We will now construct the transcription factor binding site (TFBS) matrix for known motifs (from a database like JASPAR) in peak regions. We use the `findMotifHits` function to scan for TF motif hits. This matrix will be the predictor matrix in our regression.

As mentioned, this framework offers the flexibility to add additional predictors to compete against the TF motifs for selection. Here, we add the fraction of G+C and CpG observed/expected ratio as predictors, to ensure that selected TF motifs are not just detecting a simple trend in GC or CpG composition.

```{r predictorMatrix}
# get PFMs (vertebrate TFs from Jaspar)
pfms <- getMatrixSet(JASPAR2018, list(matrixtype = "PFM", tax_group = "vertebrates"))

# randomly sample 200 PFMs for illustration purposes (for quick runtime)
set.seed(6698)
pfms <- pfms[sample(length(pfms), size = 200)]

# convert PFMs to PWMs
pwms <- toPWM(pfms)

# get TFBS on given GRanges (peaks)
peakSeq <- getSeq(BSgenome.Mmusculus.UCSC.mm10, gr)
hits <- findMotifHits(query = pwms, subject = peakSeq, min.score = 10.0,
                      BPPARAM = BiocParallel::MulticoreParam(2))

# get TFBS matrix
TFBSmatrix <- unclass(table(factor(seqnames(hits), levels = seqlevels(hits)),
                            factor(hits$pwmname, levels = name(pwms))))
TFBSmatrix[1:6, 1:6]

# remove TF motifs with 0 binding sites in all regions
zero_TF <- colSums(TFBSmatrix) == 0
sum(zero_TF)
TFBSmatrix <- TFBSmatrix[, !zero_TF]

# calculate G+C and CpG obs/expected
fMono <- oligonucleotideFrequency(peakSeq, width = 1L, as.prob = TRUE)
fDi <- oligonucleotideFrequency(peakSeq, width = 2L, as.prob = TRUE)
fracGC <- fMono[, "G"] + fMono[, "C"]
oeCpG <- (fDi[, "CG"] + 0.01) / (fMono[, "G"] * fMono[, "C"] + 0.01)

# add GC and oeCpG to predictor matrix
TFBSmatrix <- cbind(fracGC, oeCpG, TFBSmatrix)
TFBSmatrix[1:6, 1:6]
```


## Identify important TFs

We can now run randomized lasso stability selection to identify TFs that are likely to explain the log-fold changes in accessibility. 

```{r stabSelTFs}
# # randLassoStabSel() is stochastic, so we set a seed to reproduce a parallel run
# RNGkind("L'Ecuyer-CMRG")
# set.seed(123)
# se <- randLassoStabSel(x = TFBSmatrix, y = gr$logFC_liver_vs_lung, cutoff = 0.8, mc.preschedule = TRUE, mc.set.seed = TRUE, mc.cores = 2L)

# if not running in parallel mode, it is enough to use set.seed() before using the function
# ... to ensure reproducibility (with 1 core)
set.seed(123)
se <- randLassoStabSel(x = TFBSmatrix, y = gr$logFC_liver_vs_lung, cutoff = 0.8)
se

# selected TFs
colnames(se)[se$selected]
```


The stability paths visualize how predictors get selected, decreasing regularization stringency (from left to right): 
```{r plotStabilityPaths}
plotStabilityPaths(se)
```
Each line corresponds to a predictor, and we can see the selection probabilities as a function of the regularization steps, corresponding to decreasing values for the lambda regularization parameter in lasso. The predictor (TF motif) selection happens at the last step, given the specified minimum probability.

\ 

We can also visualize the selection probabilities of the selected TF motifs, optionally multiplied by the sign of the correlation to the response vector, to know how the TF relates to the change of accessibility (`directional` parameter). Note that although one can vary the `selProbMinPlot` parameter which sets the selection probability cutoff, it is best to re-run randomized lasso stability selection with the new cutoff, as this influences other parameter choices the model uses internally. See @StabSel for more details.
```{r plotSelProbs, fig.width=15, fig.height=9}
plotSelectionProb(se, directional = TRUE)
```

Next, we visualize the correlation structure of the TFBS matrix and selected TF motifs. While the colinearity of predictors is generally an issue in regression-based approaches where variables are selected, randomized lasso stability selection generally does better at selecting correlated predictors. In practice, we still see it select predictors with correlations as high as 0.9. However, it is good to keep in mind that this can be an issue, and that predictors that are extremely correlated with each other might not end up being selected. In a bad case example, if we have two identical predictors, neither of them could end up being selected if half of the times the regressions are done, one would be randomly selected, and the other one in the other half, giving each a selection probability of 0.5. If many cases of high correlations exist or this is a concern, one may consider selecting a representative set of predictors to use. 
```{r TFBScor_selected, fig.width=10, fig.height=7}
# subset the selected TFs
sel <- colnames(se)[se$selected]
se_sub <- se[, sel]

# exclude oeCpG and fracGC
excl <- colnames(se_sub) %in% c("oeCpG", "fracGC")
se_sub <- se_sub[, !excl]

# correlation matrix 
TFBSmatrixCorSel <- cor(TFBSmatrix[, colnames(se_sub)], method = "pearson")

# heatmap
pfmsSel <- pfms[match(colnames(TFBSmatrixCorSel), name(pfms))]
maxwidth <- max(sapply(TFBSTools::Matrix(pfmsSel), ncol))
seqlogoGrobs <- lapply(pfmsSel, seqLogoGrob, xmax = maxwidth)

hmSeqlogo <- rowAnnotation(logo = anno_seqlogo(seqlogoGrobs, which = "row"),
                           annotation_width = unit(2, "inch"), 
                           show_annotation_name = FALSE
)

colAnn <- HeatmapAnnotation(AUC = se_sub$selAUC, selProb = se_sub$selProb,
                            show_legend = TRUE, 
                            show_annotation_name = TRUE,
                            col = list(
                              AUC = colorRamp2(c(0, 1), c("white", "brown")),
                              selProb = colorRamp2(c(0, 1), c("white", "steelblue")))
)

Heatmap(TFBSmatrixCorSel, 
        show_row_names = TRUE, 
        show_column_names = TRUE, 
        name = "Pear. Cor.", column_title = "Selected TFs",
        col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red")), 
        right_annotation = hmSeqlogo,
        top_annotation = colAnn)
```


```{r TFBScor_all}
# exclude oeCpG and fracGC
excl <- colnames(se) %in% c("oeCpG", "fracGC")
se_sub <- se[, !excl]

# correlation matrix 
TFBSmatrixCorSel <- cor(TFBSmatrix[, colnames(se_sub)], method = "pearson")

# heatmap annotations
ann <- HeatmapAnnotation(selected = as.character(se_sub$selected), 
                         AUC = se_sub$selAUC, 
                         selProb = se_sub$selProb, 
                         col = list(selected = c("TRUE"="forestgreen", "FALSE"="grey"), 
                                    AUC = colorRamp2(c(0, 1), c("white", "brown")), 
                                    selProb = colorRamp2(c(0, 1), c("white", "steelblue")))
                         )

# heatmap
Heatmap(TFBSmatrixCorSel, show_row_names = FALSE, show_column_names = FALSE, 
        name = "Pear. Cor.", column_title = "All TFs",
        col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red")), 
        top_annotation = ann)
```

\  

We can examine the peaks that have hits for a selected TF motif of interest, ordered by absolute accessibility changes.
```{r topPeaks}
TF <- sel[2]
TF

i <- which(assay(se, "x")[, TF] > 0) # peaks that contain TF hits...
nm <- names(sort(abs(gr$logFC_liver_vs_lung[i]), decreasing = TRUE)) # ... order by |logFC|

head(gr[nm])
```


# Session info and logo
The monaLisa logo uses a drawing that was obtained from http://vectorish.com/lisa-simpson.html
under the Creative Commons attribution - non-commercial 3.0 license: https://creativecommons.org/licenses/by-nc/3.0/.

This vignette was built using:  
```{r, session}
sessionInfo()
```

# References
